clippy_lints/src/redundant_test_prefix.rs

```diff
@@ -0,0 +1,168 @@
+        if !ident.as_str().starts_with("test_") {
+            return;
+        }
+
+        if is_test_function(cx.tcx, cx.tcx.local_def_id_to_hir_id(fn_def_id)) {
+            let msg = "redundant `test_` prefix in test function name";
+            let mut non_prefixed = ident.as_str().trim_start_matches("test_").to_string();
+
+            if is_invalid_ident(&non_prefixed) {
+                // If the prefix-trimmed name is not a valid function name, do not provide an
+                // automatic fix, just suggest renaming the function.
+                span_lint_and_help(
+                    cx,
+                    REDUNDANT_TEST_PREFIX,
+                    ident.span,
+                    msg,
+                    None,
+                    "consider function renaming (just removing `test_` prefix will produce invalid function name)",
+                );
+            } else if name_conflicts(cx, body, &non_prefixed) {
+                // If `non_prefixed` conflicts with another function in the same module/scope,
+                // do not provide an automatic fix, but still emit a fix suggestion.
+                non_prefixed.push_str("_works");
+                span_lint_and_sugg(
+                    cx,
+                    REDUNDANT_TEST_PREFIX,
+                    ident.span,
+                    msg,
+                    "consider function renaming (just removing `test_` prefix will cause a name conflict)",
+                    non_prefixed,
+                    Applicability::MaybeIncorrect,
+                );
+            } else {
+                // If `non_prefixed` is a valid identifier and does not conflict with another function,
+                // so we can suggest an auto-fix.
+                span_lint_and_sugg(
+                    cx,
+                    REDUNDANT_TEST_PREFIX,
+                    ident.span,
+                    msg,
+                    "consider removing the `test_` prefix",
+                    non_prefixed,
+                    Applicability::MaybeIncorrect,
+                );
+            }
+        }
+    }
+}
+
+/// Checks whether removal of the `_test` prefix from the function name will cause a name conflict.
+///
+/// There should be no other function with the same name in the same module/scope. Also, there
+/// should not be any function call with the same name within the body of the function, to avoid
+/// recursion.
+fn name_conflicts<'tcx>(cx: &LateContext<'tcx>, body: &'tcx Body<'_>, fn_name: &str) -> bool {
+    let tcx = cx.tcx;
+    let id = body.id().hir_id;
+
+    // Iterate over items in the same module/scope
+    let (module, _module_span, _module_hir) = tcx.hir_get_module(tcx.parent_module(id));
+    for item in module.item_ids {
+        let item = tcx.hir_item(*item);
+        if let hir::ItemKind::Fn { ident, .. } = item.kind
+            && ident.name.as_str() == fn_name
+        {
+            // Name conflict found
+            return true;
+        }
+    }
```

Comment:
Nitpick: this looks like a manual implementation of `Iterator::any()`. You might want to use it directly, or not, this is a matter of taste. This could look like that once `fn_name` is interned:

```suggestion
    if module
        .item_ids
        .iter()
        .any(|item| matches!(tcx.hir_item(*item).kind, hir::ItemKind::Fn { ident, .. } if ident.name == fn_name))
    {
        return true;
    }
```